### ISSUE TITLE
kotlin imports

### ISSUE THREAD

**Issue by viral-p** [redhat-developer/vscode-java#531]
In a project, there are some interfaces set up in Kotlin, also config files, that are imported into the main Java file. These aren't recognized as valid imports, support for this would be useful. Attempting to move out of IntelliJ to VS Code, this is the only problem with the project.

**Comment by fbricon** [comment]
While I can appreciate the problem, Kotlin support is kinda out of scope of our Java support at the moment. We'll need to assess if/how we can help improve the situation. It might come from a companion extension eventually.

**Comment by fwcd** [comment]
Currently developing a [language server and an extension for Kotlin](https://github.com/fwcd/kotlin-language-server), I have [thought about this](https://github.com/fwcd/kotlin-language-server/issues/4).

In a mixed Kotlin-Java Gradle project both languages do recognize dependencies properly. They fail when trying to use classes from the other language though.

An approach to this problem might be to add compiled JVM classes located in the other languages build directory to the classpath.

**Comment by mrjones2014** [comment]
Any updates on this?

**Comment by fwcd** [comment]
@mrjones2014 Not yet as I do not have enough spare time currently. [You might want to consider contributing to the project](https://github.com/fwcd/KotlinLanguageServer#getting-started).

**Comment by coffeepac** [comment]
@fwcd do you have any suggestions on how to implement this either here or in the kotlin language server?  you mentioned copying the classfiles around but I'm not sure how to do that manually, let alone code it up.

**Comment by mrjones2014** [comment]
@fwcd If I wanted to make a pull request to your Kotlin Language Server project for this, any suggestions on where to start? How does it know (as of right now) where to look for compiled files?

**Comment by icelove82** [comment]
Can't wait for using kotlin in VSCode. When can we use kotlin in VSCode?

**Comment by Krypton17** [comment]
Any updates or news?

**Comment by hitoshisatow** [comment]
This also seems like the last piece for me as well to make VS Code viable for development. I am unable to run Java Junit classes that reference a Kotlin class as it is not able to find the class.

**Comment by robtirta** [comment]
Yes this is pretty hard since most projects nowadays mix kotlin packages in the import...

**Comment by daplf** [comment]
Hey, just in case anyone might have missed it, I built a prototype companion extension that attempts to solve this issue: https://github.com/fwcd/vscode-java-kotlin
We haven't released it yet, but if you really need this you can build it yourself for now.

**Comment by major7x** [comment]
Any updates on this issue?

**Comment by mikailcolak** [comment]
Any updates on this?

**Comment by gcocevar** [comment]
I follow the others here: are there any updates?
I'm having the same problem and it's frustrating over and over again.
The program compiles correctly but since the import of Kotlin classes into Java is not supported any reference to a Kotlin class causes many error messages in the error panel, making it very difficult to work.
Please can someone consider trying to fix this? Unfortunately I don't have too much knowledge to help more.

**Comment by linsaftw** [comment]
Any updates? It's frustrating to have to use IntelliJ for some projects.

**Comment by daplf** [comment]
Hey everyone. I finally published the companion extension I linked above to make this work: https://marketplace.visualstudio.com/items?itemName=daplf.vscode-java-kotlin

Keep in mind that this is only a prototype for now and I haven't used it with very large projects, but it seems to work well for small ones at least. If you have any problems with the extension, please create an issue. I'm happy to improve it :)

**Comment by FelixGSE** [comment]
Any updates here? Will this feature come at some point?

**Comment by dimaslanjaka** [comment]
info update

**Comment by Kangmo** [comment]
this is necessary, as we mix kotlin and java

**Comment by Clay-Ferguson** [comment]
Every big project (including mine) that is going to move from Java to Kotlin will do so incrementally rather than as a total rewrite so the need for having Kotlin recognized in in the VSCode IDE is a pretty big deal breaker, and will stop people from adopting Kotlin.

**Comment by crummy** [comment]
> Every big project (including mine) that is going to move from Java to Kotlin will do so incrementally rather than as a total rewrite so the need for having Kotlin recognized in in the VSCode IDE is a pretty big deal breaker, and will stop people from adopting Kotlin.

Agreed, but I suggest telling that to Jetbrains.

**Comment by Clay-Ferguson** [comment]
> Agreed, but I suggest telling that to Jetbrains.

Jetbrains benefits HUGELY from leaving this flaw unfixed, because it's literally the only thing stopping Kotlin & Java from being mixed in VSCode projects, and forces people to stay on IntelliJ.

**Comment by ArmoredPony** [comment]
can compiled `.class` files be imported instead of kotlin source code?

**Comment by rattrayalex** [comment]
@fbricon @fwcd this issue prevents `vscode-java` users from being able to see docs-on-hover or jump-to-definition from the `openai` and `anthropic` Java SDKs, which are implemented in Kotlin. 

Would you be open to sponsorship or know of a trusted contributor we could sponsor to move this forward? (email in profile)

**Comment by fwcd** [comment]
I don't really have the capacity to maintain the language server more actively these days, but feel free to ask some of the other contributors. @daplf wrote an extension for this a while ago: https://github.com/daplf/vscode-java-kotlin

Would love to see better Kotlin/Java interop in VSCode too.

**Comment by TomerAberbach** [comment]
@fwcd do you know if this issue is the same or different from importing a Kotlin _library_ into Java? Versus importing a Kotlin file in the same project into Java?

I ask because it sounds like in the latter case the import is completely unresolved while in the former case the import is resolved, but you don't get docs-on-hover or jump-to-definition.

**Comment by rattrayalex** [comment]
Thanks @fwcd ! Do you have a sense of whether that extension could feasibly be merged into this one? 

Ideally Java users would not have to install a Kotlin-related extension just to use a library that happens to be built with Kotlin.

**Comment by fwcd** [comment]
@TomerAberbach It's been a while since I've dug into the classpath logic so I'm probably not the right person to ask 😄 That said, I remember it being a bit fiddly, since it very much depends on how your project is built and structured. The idea was definitely to support mixed Java/Kotlin codebases, not sure how far we got with that.

**Comment by daplf** [comment]
@rattrayalex @TomerAberbach Hey 👋 

Your use case is one I hadn't really considered and it's not really what my extension was made for. However, the issue is fairly similar.

In a nutshell, my extension allows Kotlin source files used in mixed Java/Kotlin codebases to be added to the project classpath, so that the Java Language Server doesn't complaint about missing symbols (defined in Kotlin).

However, my extension merely adds the compiled bytecode to the project classpath in JDT LS. This means we have the same problem. Any additional language support, such as `jump-to-definition` or `docs-on-hover` is missing, because JDT LS only knows about the bytecode (and even if it knew about the source files, it wouldn't be able to understand them, since they are written in Kotlin).

The solution to this will likely be something along these lines: https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/2583#issuecomment-1507832846

So there needs to be a companion extension that advertises Java language support and is thus able to provide hovers and definitions for Kotlin symbols used in Java source files. This extension would likely need to connect to some other language server that is capable of rendering docs for the Kotlin symbols and provide definition results. Since the actual project is managed by JDT LS, this separate language server would potentially need to make requests to JDT LS (via the `delegateCommandHandler` pattern) to fetch any information necessary to resolve the symbols.

I don't have a lot of experience with JDT LS and their extensions apart from the one I built, so I'm not entirely sure this would actually work, but it seems to me like this is how some similar use cases are currently being done (e.g., `vscode-microprofile`, see https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/1731)

Regarding upstreaming any of this to `vscode-java`, I can't speak on behalf of the extension maintainers, but as far as I understand, this extension is meant only for Java support. Kotlin (in any shape or form) is out of scope. Therefore, I suspect you will always require a companion extension.

**Comment by rattrayalex** [comment]
Thanks for the note @daplf !

>  I can't speak on behalf of the extension maintainers, but as far as I understand, this extension is meant only for Java support. Kotlin (in any shape or form) is out of scope.

I am curious whether maintainers share this view. Should a Java developer installing a package like `openai-java` get intellisense if that package is written with (some) Kotlin?

**Comment by fbricon** [comment]
So @snjeza came up with something that might help here (#4313 and https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/3682), and we need to check if it works on real life projects. Can anyone share a project that mixes java and kotlin sources?

**Comment by fbricon** [comment]
So we merged some partial support for Kotlin, thanks to @snjeza's work. Java files now see kotlin classes.  
It comes with a bunch of caveats though:

- it's only supported with Gradle projects imported via the Buildship plugin (i.e not via vscode-gradle, or any Maven projects)
- It does not provide any Kotlin intelligence (no completion, hover, references ... in kotlin files). Though kotlin errors are reported in .kt files
- it's fairly slow on large projects ( a gradle build is triggered under the hood when a kotlin file is saved)
- Navigating to kotlin classes open the decompiled kotlin file, instead of the actual source one. We might be able to fix that though (#4345).

**Comment by Emily1799** [comment]
>it's only supported with Gradle projects imported via the Buildship plugin (i.e not via vscode-gradle, or any Maven projects)

What would be involved with removing this limitation? Is it a total rethinking, or just something like a few more steps, or support in vscode-gradle?
